### PR TITLE
rgw_sal_motr :[CORTX-31533] Fix delete-object API for unversioned object

### PR DESCRIPTION
**Pull request by shriya-deshmukh**
Problem:

delete-object API is not working for unversioned bucket.

Solution:

Pass version-id in key name(obj1[version-id]) while deleting the object.

Signed-off-by: Shriya Deshmukh <shriya.deshmukh@seagate.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by sachitanands — APPROVED** [review]
looks good

**Review by priyanka25081999 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1658,25 +1658,41 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
...
 {
+  int rc;
+  bufferlist bl;
+  string tenant_bkt_name = get_bucket_name(source->get_bucket()->get_tenant(), source->get_bucket()->get_name());
+  string bucket_index_iname = "motr.rgw.bucket.index." + tenant_bkt_name;
+  std::string delete_key;
+  rgw_bucket_dir_entry ent;
+  rc = source->get_bucket_dir_ent(dpp, ent);
+  if (rc < 0) 
+    return rc;
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:1669]: we should log the error here

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -1658,25 +1658,41 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
...
+
   if (source->have_instance()) {
     rgw_obj_key& key = source->get_key();
+    delete_key = source->get_key().to_str();
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:1673]: std::string delete_key = key->to_str()

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1658,25 +1658,41 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
...
     if (key.have_null_instance())
-      key.instance.clear();
+    {
+      // Read null reference key to get null version obj key.
+      rc = source->fetch_null_obj_reference(dpp, delete_key);
+      if (rc < 0)
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply src/rgw/rgw_sal_motr.cc:1680]: We have already added failure logs in fetch_null_obj_reference fun.

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -1658,25 +1658,41 @@ MotrObject::MotrDeleteOp::MotrDeleteOp(MotrObject *_source, RGWObjectCtx *_rctx)
...
-  rgw_bucket_dir_entry ent;
-  int rc = source->get_bucket_dir_ent(dpp, ent);
-  if (rc < 0) {
-    return rc;
+  else{
+    if (ent.key.instance == "null")
+    {
+      rc = source->fetch_null_obj_reference(dpp, delete_key);
+      if (rc < 0)
+        return rc;
```
[the remark this answers is not in the document]
Comment (shriya-deshmukh) [reply src/rgw/rgw_sal_motr.cc:1689]: We have already added failure logs in fetch_null_obj_reference fun.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (shriya-deshmukh) [reply src/rgw/rgw_sal_motr.cc]: Done.

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (shriya-deshmukh) [reply src/rgw/rgw_sal_motr.cc:1671]: Done

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**